distributed/tests/test_failed_workers.py

```diff
@@ -48,21 +49,59 @@ def test_submit_after_failed_worker_sync(loop):
             assert total.result() == sum(map(inc, range(10)))
 
 
-@pytest.mark.slow()
-@pytest.mark.parametrize("compute_on_failed", [False, True])
-@gen_cluster(client=True, config={"distributed.comm.timeouts.connect": "500ms"})
-async def test_submit_after_failed_worker_async(c, s, a, b, compute_on_failed):
-    async with Nanny(s.address, nthreads=2) as n:
-        await c.wait_for_workers(3)
-
-        L = c.map(inc, range(10))
-        await wait(L)
+@pytest.mark.parametrize("when", ["closing", "closed"])
+@pytest.mark.parametrize("y_on_failed", [False, True])
+@pytest.mark.parametrize("x_on_failed", [False, True])
+@gen_cluster(
+    client=True,
+    nthreads=[("", 1)] * 2,
+    config={"distributed.comm.timeouts.connect": "1s"},
+)
+async def test_submit_after_failed_worker_async(
+    c, s, a, b, x_on_failed, y_on_failed, when, monkeypatch
+):
+    a_ws = s.workers[a.address]
+
+    L = c.map(
+        inc,
+        range(10),
+        workers=[b.address if x_on_failed else a.address],
+        allow_other_workers=True,
+    )
+    await wait(L)
 
-        kill_task = asyncio.create_task(n.kill())
-        compute_addr = n.worker_address if compute_on_failed else a.address
-        total = c.submit(sum, L, workers=[compute_addr], allow_other_workers=True)
-        assert await total == sum(range(1, 11))
-        await kill_task
+    if when == "closed":
+        await b.close()
+        await async_poll_for(lambda: b.address not in s.workers, timeout=5)
+    elif when == "closing":
+        orig_remove_worker = s.remove_worker
+        in_remove_worker = asyncio.Event()
+        wait_remove_worker = asyncio.Event()
+
+        async def remove_worker(*args, **kwargs):
+            in_remove_worker.set()
+            await wait_remove_worker.wait()
+            return await orig_remove_worker(*args, **kwargs)
+
+        monkeypatch.setattr(s, "remove_worker", remove_worker)
+        await b.close()
+        await in_remove_worker.wait()
+        assert s.workers[b.address].status.name == "closing"
+
+    total = c.submit(
+        sum,
+        L,
+        key="y",
+        workers=[b.address if y_on_failed else a.address],
+        allow_other_workers=True,
+    )
+
+    await wait_for_state("y", "processing", s, interval=0)
+    assert s.tasks["y"].processing_on is a_ws
```

Comment:
I feel this is a bit too specific. We don't _actually_ care about this specific state too much.

```suggestion
    while "y" not in s.tasks:
        await asyncio.sleep(0.05)
```

does the trick as well and properly triggers the assert of https://github.com/dask/distributed/issues/8019

or even better without any sleeping

```python

    def blocked_sum(arg, event):
        event.set()
        return sum(arg)

    event = Event()
    total = c.submit(
        blocked_sum,
        L,
        key="y",
        event=event,
        workers=[b.address if y_on_failed else a.address],
        allow_other_workers=True,
    )

    await event.wait()
```

I'm typically discouraging usage of `wait_for_state`, particularly with very low intervals since this can actually flood the event loop and cause weird issues like the following https://github.com/dask/distributed/issues/6080